src/VisualStudio/Core/Def/Diagnostics/VisualStudioDiagnosticAnalyzerService.cs

```diff
@@ -298,154 +161,156 @@ private void OnRunCodeAnalysisForSelectedProjectStatus(object sender, EventArgs
     private void OnRunCodeAnalysisForSelectedProject(object sender, EventArgs args)
     {
         if (VisualStudioCommandHandlerHelpers.TryGetSelectedProjectHierarchy(_serviceProvider, out var hierarchy))
-        {
             RunAnalyzers(hierarchy);
-        }
     }
 
     public void RunAnalyzers(IVsHierarchy? hierarchy)
     {
-        var project = GetProject(hierarchy);
+        // If a new command comes in to run analyzers again, cancel any existing operation in progress and start a new one.
+        var cancellationToken = _cancellationSeries.CreateNext();
+
         var solution = _workspace.CurrentSolution;
-        var projectOrSolutionName = project?.Name ?? PathUtilities.GetFileName(solution.FilePath);
+        var project = GetProject(solution, hierarchy);
 
         // Handle multi-tfm projects - we want to run code analysis for all tfm flavors of the project.
-        ImmutableArray<Project> otherProjectsForMultiTfmProject;
+        string progressName;
+        ImmutableArray<Project> projectsToAnalyze;
         if (project != null)
         {
-            otherProjectsForMultiTfmProject = [.. solution.Projects.Where(
-                p => p != project && p.FilePath == project.FilePath && p.State.NameAndFlavor.name == project.State.NameAndFlavor.name)];
-            if (!otherProjectsForMultiTfmProject.IsEmpty)
-                projectOrSolutionName = project.State.NameAndFlavor.name;
+            projectsToAnalyze = solution.Projects.WhereAsArray(
+                static (otherProject, project) => otherProject.FilePath == project.FilePath && otherProject.State.NameAndFlavor.name == project.State.NameAndFlavor.name,
+                project);
+
+            progressName = projectsToAnalyze.Length == 1
+                ? project.Name
+                : project.State.NameAndFlavor.name ?? project.Name;
         }
         else
         {
-            otherProjectsForMultiTfmProject = [];
+            projectsToAnalyze = [.. solution.Projects];
+            progressName = PathUtilities.GetFileName(solution.FilePath) ?? FeaturesResources.Solution;
         }
```

Comment:
simplified the logic for determining what to run.  you either run for the solution (all projects), or for one project (which then pulls in the related TFMs for it).